src/js/core.js

```diff
@@ -0,0 +1,616 @@
+     * @property {String} name - The name of the device associated with the port.
+     * @property {String} manufacturer - The manufacturer of the device associated with the port.
+     * @property {String} type - The type of port, either "input", or "output".
+     * @property {String} state - The state of the port, either "open", "pending", or "closed".
+     * @property {String} connected - "connected" if the port is connected, "disconnected" otherwise.
+     */
+
+    /**
+     * @typedef CallRegister
+     * @property {Array<Array>} send - An array of the arguments passed when called the mock's `send` method.
+     * @property {Array<Array>} clear - An array of the arguments passed when called the mock's `clear` method.
+     * @property {Array<Array>} open - An array of the arguments passed when called the mock's `opeen` method.
+     * @property {Array<Array>} close - An array of the arguments passed when called the mock's `close` method.
+     */
+
+    /**
+     * @typedef MIDIPort
+     * @type {Object}
+     * @property {String} id - The unique ID of the device associated with the port.
+     * @property {String} name - The name of the device associated with the port.
+     * @property {String} manufacturer - The manufacturer of the device associated with the port.
+     * @property {String} type - The type of port, either "input", or "output".
+     * @property {String} state - The state of the port, either "open", "pending", or "closed".
+     * @property {String} connected - "connected" if the port is connected, "disconnected" otherwise.
+     * @property {Function} send - A function that handles requests to send messages to this port.
+     * @property {Function} clear - A function that handles requests to clear messages in progress being sent by this port.
+     * @property {Function} open - A function that will handle requests to open a connection to this port.
+     * @property {Function} close - A function that will handle requests to close the connection to this port.
+     */
+
+    /** @typedef MIDIInput
+     * @type {MIDIPort}
+     * @property {AddEventListener} addEventListener - Add a listener for events the input receives.
+     * @property {DispatchEvent} dispatchEvent - Dispatch an event to this input.
+     */
+
+    /**
+     * @typedef MIDIOutput
+     * @type {MIDIPort}
+     * @property {Function} send - A function that handles requests to send messages to this port.
+     * @property {Function} clear - A function that handles requests to clear messages in progress being sent by this port.
+     */
+
+    /**
+     * An array of data representing a MIDI message, for full details, see:
+     * https://www.midi.org/specifications-old/item/table-1-summary-of-midi-message
+     *
+     * @typedef MIDIMessageData
+     * @type {Uint8Array}
+     * @return {Promise} - A promise that will resolve when access is granted or reject otherwise.
```

Comment:
Should explain what it will resolve to - looks like it is a native MIDIAccess object?

Reply:
Yes, I added a comment to that effect.